DataServer/devices.py

```diff
@@ -30,7 +32,10 @@ class Device:
     fw_version: str = ""
     location: str = ""
     location_units: str = ""
-    _last_comms: Optional[dt.datetime] = None
+    last_comms: Optional[dt.datetime] = None
+    offline: bool = True
+    on_timeout: Optional[Callable[[devices.Device], None]] = None
+    on_connect: Optional[Callable[[devices.Device], None]] = None
```

Comment:
Is it possible that there might be multiple callbacks?

You may want to generalize the eventing mechanism:
```
class Event(enum.Enum):
    EVENT1 = enum.auto()
    EVENT2 = enum.auto()
    ...

@dataclasses.dataclass
class Device:
    def __post_init__(self):
        self.__event_cb: typing.Dict[Event, typing.Set[int]] = {evt:set() for evt in Event}
        self.__cbs: typing.Dict[int, typing.Callable] = {}
        ...

    def register_cb(self, event: Event, cb_: Callable) -> int:
        fn_id = id(cb_)
        self.__event_cb[event].add(fn_id)
        self.__cbs[fn_id] = cb_
        return fn_id

    def unregister_cb(self, fn_id: int) -> None:
        for cb_list in self.__event_cb.values():
            if fn_id in cb_list:
                cb_list.remove(fn_id)
        if fn_id in self.__cbs:
            self.__cbs.pop(fn_id)

    def do_event(self, event: Event, ...) -> None:
        for cb_ in self.__event_cb[event]:
            cb_(...)
```